pints/_mcmc/_dual_averaging.py

```diff
@@ -0,0 +1,241 @@
+        return self._inv_mass_matrix
+
+    def set_inv_mass_matrix(self, inv_mass_matrix):
+        """
+        We calculate the mass matrix whenever the inverse mass matrix is set
+        """
+        if inv_mass_matrix.ndim == 1:
+            self._mass_matrix = 1.0 / inv_mass_matrix
+            self._inv_mass_matrix = inv_mass_matrix
+        else:
+            try:
+                self._mass_matrix = np.linalg.inv(inv_mass_matrix)
+            except np.linalg.LinAlgError:
+                print('WARNING: adapted mass matrix is ill-conditioned')
+                return
+            self._inv_mass_matrix = inv_mass_matrix
+
+    def get_mass_matrix(self):
+        """ return the mass matrix """
+        return self._mass_matrix
+
+    def get_epsilon(self):
+        """ return the step size """
+        return self._epsilon
+
+    def step(self, x, accept_prob):
+        """
+        Perform a single step of the adaption
+
+        Arguments
+        ---------
+
+        x: ndarray
+            the next accepted mcmc parameter point
+
+        accept_prob: float
+            the acceptance probability of the last NUTS/HMC mcmc step
+        """
+
+        if not self._adapting:
+            return
+
+        self._counter += 1
+
+        if self._counter >= self._warmup_steps:
+            self._epsilon = self.final_epsilon()
+            self._adapting = False
+            return False
+
+        self.adapt_epsilon(accept_prob)
+        if self._counter > self._initial_window:
+            self.add_parameter_sample(x)
+
+        if self._counter >= self._next_window:
+            self.set_inv_mass_matrix(self.calculate_sample_variance())
+            if self._counter >= self._warmup_steps - self._terminal_window:
+                self._next_window = self._warmup_steps
+            else:
+                self._base_window *= 2
+                self._next_window = min(
+                    self._counter + self._base_window,
+                    self._warmup_steps - self._terminal_window
+                )
+            self.init_sample_covariance(self._next_window - self._counter)
+            return True
+            #self._epsilon = self.final_epsilon()
+
+        return False
+
+    def init_adapt_epsilon(self, epsilon):
+        """
+        Start a new dual averaging adaption for epsilon
+        """
+        self._epsilon = epsilon
+        self._mu = np.log(10 * self._epsilon)
+        self._log_epsilon_bar = np.log(1)
+        self._h_bar = 0.0
+        self._adapt_epsilon_counter = 0
+
+    def adapt_epsilon(self, accept_prob):
+        """
+        Perform a single step of the dual averaging scheme
+        """
+
+        if accept_prob > 1:
+            accept_prob = 1.0
+
+        self._adapt_epsilon_counter += 1
+        counter = self._adapt_epsilon_counter
+
+        eta = 1.0 / (counter + self._t0)
+
+        self._h_bar = (1 - eta) * self._h_bar \
+            + eta * (self._target_accept_prob - accept_prob)
```

Comment:
This is the line I was unsure about when comparing to Algorithm 6 in Hoffman.

Reply:
the acceptance probability is calculated differently for multinomial sampling, so this particular class generalises the iteration step to use ``accept_prob``, which is given as an argument to `step`

Reply:
Cool, thanks!